Modules/_testcapi/long.c

```diff
@@ -141,13 +141,177 @@ pylong_aspid(PyObject *module, PyObject *arg)
 }
 
 
+static PyObject *
+layout_to_dict(const PyLongLayout *layout)
+{
+    PyObject *dict = PyDict_New();
+    if (dict == NULL) {
+        goto error;
+    }
+
+    PyObject *value = PyLong_FromUnsignedLong(layout->bits_per_digit);
+    if (value == NULL) {
+        goto error;
+    }
+    int res = PyDict_SetItemString(dict, "bits_per_digit", value);
+    Py_DECREF(value);
+    if (res < 0) {
+        goto error;
+    }
+
+    value = PyLong_FromUnsignedLong(layout->digit_size);
+    if (value == NULL) {
+        goto error;
+    }
+    res = PyDict_SetItemString(dict, "digit_size", value);
+    Py_DECREF(value);
+    if (res < 0) {
+        goto error;
+    }
+
+    value = PyLong_FromLong(layout->digits_order);
+    if (value == NULL) {
+        goto error;
+    }
+    res = PyDict_SetItemString(dict, "digits_order", value);
+    Py_DECREF(value);
+    if (res < 0) {
+        goto error;
+    }
+
+    value = PyLong_FromLong(layout->endianness);
+    if (value == NULL) {
+        goto error;
+    }
+    res = PyDict_SetItemString(dict, "endianness", value);
+    Py_DECREF(value);
+    if (res < 0) {
+        goto error;
+    }
+
+    return dict;
+
+error:
+    Py_XDECREF(dict);
+    return NULL;
+}
+
+
+static PyObject *
+pylong_export(PyObject *module, PyObject *obj)
+{
+    PyLongExport export_long;
+    if (PyLong_Export(obj, &export_long) < 0) {
+        return NULL;
+    }
+
+    if (export_long.digits == NULL) {
+        return PyLong_FromInt64(export_long.value);
+        // PyLong_FreeExport() is not needed in this case
+    }
+
+    assert(PyLong_GetNativeLayout()->digit_size == sizeof(digit));
+    const digit *export_long_digits = export_long.digits;
+
+    PyObject *digits = PyList_New(0);
+    for (Py_ssize_t i=0; i < export_long.ndigits; i++) {
+        PyObject *item = PyLong_FromUnsignedLong(export_long_digits[i]);
+        if (item == NULL) {
+            goto error;
+        }
+
+        if (PyList_Append(digits, item) < 0) {
+            Py_DECREF(item);
+            goto error;
+        }
+        Py_DECREF(item);
+    }
+
+    PyObject *res = Py_BuildValue("(iN)", export_long.negative, digits);
+
+    PyLong_FreeExport(&export_long);
+    assert(export_long._reserved == 0);
+
+    return res;
+
+error:
+    Py_DECREF(digits);
+    PyLong_FreeExport(&export_long);
+    return NULL;
+}
+
+
+static PyObject *
+pylongwriter_create(PyObject *module, PyObject *args)
+{
+    int negative;
+    PyObject *list;
+    if (!PyArg_ParseTuple(args, "iO!",
+                          &negative,
+                          &PyList_Type, &list))
+    {
+        return NULL;
+    }
+    Py_ssize_t ndigits = PyList_GET_SIZE(list);
+
+    digit *digits = PyMem_Malloc((size_t)ndigits * sizeof(digit));
+    if (digits == NULL) {
+        PyErr_NoMemory();
+        return NULL;
```

Comment:
`PyErr_NoMemory` always returns `NULL`

```suggestion
        return PyErr_NoMemory();
```